suite/tests/client-interface/stolen-reg.dll.c

```diff
@@ -0,0 +1,138 @@
+/* **********************************************************
+ * Copyright (c) 2020 Google, Inc.  All rights reserved.
+ * **********************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/*
+ * API regression test for stolen register translation.
+ */
+
+#include "dr_api.h"
+#include "client_tools.h"
+
+#define BAD_VALUE 0xdeadbeef
+
+/* We assume the app is single-threaded and don't worry about races. */
+static ptr_int_t app_stolen_reg_val;
+
+static void
+restore_event(void *drcontext, void *tag, dr_mcontext_t *mcontext, bool restore_memory,
+              bool app_code_consistent)
+{
+    dr_log(drcontext, DR_LOG_ALL, 2, "Changing the stolen reg value from %ld to %ld\n",
+           mcontext->IF_ARM_ELSE(r10, r28), app_stolen_reg_val);
+    mcontext->IF_ARM_ELSE(r10, r28) = app_stolen_reg_val;
+}
+
+static void
+do_flush(app_pc next_pc)
+{
+    dr_fprintf(STDERR, "Performing synchall flush\n");
+    if (!dr_flush_region(NULL, ~0UL))
+        DR_ASSERT(false);
+    void *drcontext = dr_get_current_drcontext();
+    dr_mcontext_t mcontext;
+    mcontext.size = sizeof(mcontext);
+    mcontext.flags = DR_MC_ALL;
+    dr_get_mcontext(drcontext, &mcontext);
+    mcontext.pc = dr_app_pc_as_jump_target(dr_get_isa_mode(drcontext), next_pc);
```

Comment:
PR #4507 has a bug where it fails to decorate the PC in a similar redirect for the new flush in tracer.cpp.  I'll submit a separate PR to fix that.

Reply:
I suppose you  mean PR #4491?

Reply:
I filed #4527 on this.

Reply:
Woah, somehow I missed your pointer to #4491 -- yes, correct.  I'll update PR #4528 on merge.